### PR TITLE
Run unit tests only for plugins with changes

### LINKED ISSUE #1792: Unit testing should be skipped for plugins not related to changes in a pull request
## Feature Description

When opening a pull request that changes files to one plugin, only the unit tests for that plugin should run.  This will greatly speed up our workflows (and reduce wasted compute), especially now that code coverage is being performed. That said, would it be problematic if PRs didn't include coverage reports for all plugins? 

Note that a PR that modifies Optimization Detective should also cause the tests for Embed Optimizer and Image Prioritizer to also run, since they are dependent on that plugin.

Naturally, all tests would run for all plugins for commits added to `trunk`.

This has been suggested before by @thelovekesh, but I don't remember where.

### PR DESCRIPTION
**Pull request by ShyamGadde**
## Summary

Fixes #1792 

## Relevant technical choices

- Run unit tests only for the plugin(s) affected by the changes in the pull request.
- Changes to the Optimization Detective plugin trigger tests for Embed Optimizer and Image Prioritizer due to their dependencies.
- All tests will run for all plugins when commits are added to the `trunk` branch.

### CODE REVIEW

File: .github/workflows/php-test-plugins.yml [hunk 1]
```diff
@@ -57,6 +57,54 @@ jobs:
     steps:
       - uses: styfle/cancel-workflow-action@0.12.1
       - uses: actions/checkout@v4
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+        with:
+          dir_names: true # Output unique changed directories.
+          dir_names_max_depth: 2
```
Comment (ShyamGadde) [inline .github/workflows/php-test-plugins.yml:71]: This limits the directory output to a maximum depth of 2. For example, `plugins/<plugin-name>/tests` will be returned as `plugins/<plugin-name>`.  
Since `dir_names: true` ensures only unique directories are listed, combining it with `dir_names_max_depth: 2` allows us to extract a clean list of changed plugin names. This avoids additional logic in the next step for determining modified plugins.

File: .github/workflows/php-test-plugins.yml [hunk 2]
```diff
@@ -57,6 +57,54 @@ jobs:
...
+          )
+          for PLUGIN in "${ALL_CHANGED_PLUGINS[@]}"; do
+            if [[ -n "${PLUGIN_DEPENDENCIES[$PLUGIN]}" ]]; then
+              for DEP in ${PLUGIN_DEPENDENCIES[$PLUGIN]}; do
+                if [[ ! " ${ALL_CHANGED_PLUGINS[@]} " =~ " ${DEP} " ]]; then
+                  ALL_CHANGED_PLUGINS+=("$DEP")
+                fi
+              done
+            fi
+          done
```
Comment (ShyamGadde) [inline .github/workflows/php-test-plugins.yml]: Currently, `optimization-detective` is the only plugin with dependencies, so we could simplify this by checking for it explicitly instead of maintaining a general dependency structure. However, I think this approach keeps things flexible if more dependencies are introduced in the future.

File: .github/workflows/php-test-plugins.yml [hunk 3]
```diff
@@ -57,6 +57,54 @@ jobs:
...
+            plugins:
+              - 'plugins/**'
+            config:
+              - '.github/workflows/php-test-plugins.yml'
+              - '.wp-env.json'
+              - '**/package.json'
+              - 'package-lock.json'
+              - 'phpunit.xml.dist'
+              - 'composer.json'
+              - 'composer.lock'
```
Comment (ShyamGadde) [inline .github/workflows/php-test-plugins.yml:83]: Should we add `tools/phpunit/bootstrap.php` here along with `phpunit.xml.dist`?

Comment (westonruter) [reply .github/workflows/php-test-plugins.yml:83]: Yes, that makes sense to me. If the plugin bootstrap file changes, we should re-run tests. But what does this `files_yaml` do here?

Comment (westonruter) [reply .github/workflows/php-test-plugins.yml:83]: Looks like it results in populating `steps.changed-files.outputs`. So yeah, if the plugin bootstrap file changes than all tests should run as `steps.changed-files.outputs.config_any_changed` is used below.

File: .github/workflows/php-test-plugins.yml [hunk 4]
```diff
@@ -57,6 +57,54 @@ jobs:
...
+          done
+
+          # Define and add plugin dependencies (e.g., optimization-detective triggers others).
+          declare -A PLUGIN_DEPENDENCIES=(
+            ["optimization-detective"]="embed-optimizer image-prioritizer"
+          )
+          for PLUGIN in "${ALL_CHANGED_PLUGINS[@]}"; do
+            if [[ -n "${PLUGIN_DEPENDENCIES[$PLUGIN]}" ]]; then
+              for DEP in ${PLUGIN_DEPENDENCIES[$PLUGIN]}; do
+                if [[ ! " ${ALL_CHANGED_PLUGINS[@]} " =~ " ${DEP} " ]]; then
```
Comment (westonruter) [inline .github/workflows/php-test-plugins.yml]: PhpStorm (via Shellcheck) is flagging this line:

> Arrays implicitly concatenate in [[ ]]. Use a loop (or explicit * instead of @).
> See [SC2199](https://github.com/koalaman/shellcheck/wiki/SC2199).

File: .github/workflows/php-test-plugins.yml (hunk 3 above)
Comment (ShyamGadde) [reply .github/workflows/php-test-plugins.yml:83]: Exactly. `files_yaml` allows defining file and directory patterns in YAML to detect changes. The action then prefixes outputs with the specified key (e.g., in `steps.changed-files.outputs.config_any_changed`, the `any_changed` output is prefixed with `config_`), simplifying condition checks downstream.

File: .github/workflows/php-test-plugins.yml (hunk 4 above)
Comment (ShyamGadde) [reply .github/workflows/php-test-plugins.yml]: Fixed it in 11949ac.

File: .github/workflows/php-test-plugins.yml [hunk 5]
```diff
@@ -57,6 +59,55 @@ jobs:
...
+          fi
+
+          declare -a ALL_CHANGED_PLUGINS=()
+          for DIR in ${{ steps.changed-files.outputs.plugins_all_changed_files }}; do
+            PLUGIN_NAME=$(basename "$DIR")
+            ALL_CHANGED_PLUGINS+=("$PLUGIN_NAME")
+          done
+
+          # Define and add plugin dependencies (e.g., optimization-detective triggers others).
+          declare -A PLUGIN_DEPENDENCIES=(
```
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: Maybe call this `PLUGIN_DEPENDENTS` to clarify? In this case, Optimization Detective is a _dependency_ of the other two, but the other two are _dependents_ of Optimization Detective.

**Review by felixarntz — COMMENTED** [review]
@ShyamGadde This looks great to me so far, just one minor comment.

Before we merge this, I would like to discuss though under which conditions we should do this. To me, it makes sense to only run selected unit tests for PRs, to speed up the process. But running the whole test suite also has value, e.g. to warn us early in case something breaks somewhere else where we didn't change any files - which can happen e.g. based on WordPress Core updates.

How about we implement this change so that it only applies to pull requests, but for merges to our main branches it would still run all test suites? IMO this would be the best of both worlds:
* Running tests against PRs would be more efficient, faster and cost less resources.
* Running tests against code merged into `trunk` and release branches would still be ensured to be more comprehensively tested.

It probably wouldn't be a huge lift to change that, we could make the step that determines the changed files and plugins conditional to `pull_request` events and otherwise return the list of all plugins.

WDYT? cc @westonruter @thelovekesh

**Review by felixarntz — CHANGES_REQUESTED** [review]
Marking this as "Request changes" until we resolve the outstanding discussion points on the approach in #1792.

**Comment by westonruter** [comment]
> How about we implement this change so that it only applies to pull requests, but for merges to our main branches it would still run all test suites? IMO this would be the best of both worlds:

This should already be the case with this logic:

https://github.com/WordPress/performance/blob/af6f8ff933a733bf9764514007a72356b3d5bb79/.github/workflows/php-test-plugins.yml#L83-L87

If any of the config files are changed _or_ if the changes are not part of a PR (i.e. a push to `trunk`) then all plugins are tested.

**Comment by felixarntz** [comment]
@westonruter Thanks, I missed that. LGTM in that regard then 👍 

In that case, this leaves the outstanding point @swissspidy raised about confusing codecov reports.

**Comment by ShyamGadde** [comment]
> In that case, this leaves the outstanding point [@swissspidy](https://github.com/swissspidy) raised about confusing Codecov reports.

I think a good solution to this would be the [`carryforward` flag](https://docs.codecov.com/docs/carryforward-flags). Codecov provides this flag to allow previously uploaded coverage reports to persist across runs, meaning untested files retain their last known coverage **instead of being omitted from the report, which can lead to misleading drops in coverage**. However, this won't work in the current setup. Right now, we are using two flags, `single` and `multisite`, and partial reports are being uploaded under these flags, which results in the PR comments showing a **drop in reported code coverage** (as untested files appear as uncovered). For `carryforward` to work, I believe no reports should be uploaded using the flag being carried forward; otherwise, the newly uploaded report will override the carried-forward coverage.

Another possible approach is to [**only show patch coverage in the PR comment**](https://docs.codecov.com/docs/changing-your-pr-comment-format), as that isn’t affected by partial testing. This would be the simpler solution, ensuring that PR reports remain clear and focused on the actual changes. Since full test runs still occur on merge commits and commits to `trunk`, the overall project coverage will still be recalculated periodically.

That said, the first approach (using `carryforward`) **might still work with some adjustments**. Instead of using just `single` and `multisite`, we could assign **unique flags per plugin** so that their coverage data is carried forward independently. That way, when a report for a particular plugin isn’t uploaded, its previous report would get carried forward automatically. (I'm still reviewing Codecov's documentation to confirm if this setup would work as expected.)

### Screenshots:

1. **With partial testing:**  
    image
2. **PR comment with only patch coverage:**  
    image

cc @felixarntz @westonruter

File: .github/workflows/php-test-plugins.yml (hunk 5 above)
Comment (ShyamGadde) [reply .github/workflows/php-test-plugins.yml]: That makes sense. Updated it in 90a5ab7.

**Comment by ShyamGadde** [comment]
So, after experimenting with Codecov's coverage reporting and reviewing its documentation, I’ve concluded that if we want to continue with partial testing, there are two possible ways to handle Codecov’s potentially confusing reports—depending on whether we want to display the project's overall coverage in the PR comment or not.

#### **Option 1: Show Project Coverage in PR Comment**

If we want to **display project coverage** in the PR comment, we need to use the `carryforward` flag to handle missing reports. But that would mean that we will not be able to use the flags such as `single` and `multisite` as we will be uploading partial reports with these flags when we don't test all the plugins which will lead to the confusing reports.

To address this, we can:

- Remove the `single` and `multisite` flags (also delete them from the Codecov dashboard).
- Instead, introduce **plugin-specific flags** and group single-site and multisite reports under those flags.

That way when we don't test a particular plugin the result will look something like in the below screenshot and even the project coverage which is displayed won't be affected when partial testing.

image

I personally find the reports (for both single and multisite) grouped by plugin flags **more intuitive to read** (because of how Codecov [merges these reports](https://docs.codecov.com/docs/merging-reports)) than the old reporting format, where we see `38.45%` for `single` and `65.86%` for `multisite` in PR comments as seen below:

image

#### **Option 2: Show Only Patch Coverage in PR Comment**

A simpler alternative is to **only show patch coverage** in PR comments. This won't show project-wide coverage fluctuations when running partial tests and keeps the report focused on the actual changes.

Additionally, if we were to only show **"Patch coverage"**, we could:

- Limit Codecov's **"Project Coverage"** status check to **only commits to the `trunk` branch**, preventing confusing reports in Codecov's dashboard and GitHub's Actions runs. Since we run all tests on commits to `trunk`, everything should function normally there.
- **Consider uploading reports without the `single` and `multisite` flags**, since they are optional. This would prevent temporary drops in coverage reporting under these flags.

With this approach, the PR comment will look like this:

image

---

Given these options, assuming we still want to do partial testing, which approach would be preferable?

cc @felixarntz @westonruter

**Comment by westonruter** [comment]
@ShyamGadde Option 2 sounds like the better approach to me! Thank you for doing that research.

File: codecov.yml [hunk 6]
```diff
@@ -8,8 +8,13 @@ coverage:
         target: auto
         threshold: 80%
         base: auto
+        branches:
+          - trunk
         informational: true
     patch:
       default:
         threshold: 80%
```
Comment (ShyamGadde) [inline codecov.yml]: I noticed that the `threshold` for patch coverage is set to `80%`, which (based on [Codecov’s docs](https://docs.codecov.com/docs/commit-status#threshold)) means that coverage can drop by up to `80%` while still passing. I'm wondering if this was intentional—perhaps it was meant to ensure a minimum of `80%` coverage rather than allowing such a large drop?

Additionally, since `target` isn't explicitly set, it defaults to the project coverage from the pull request base. That results in PR checks displaying something like:

image

Would it make sense to set `target: 100%` for patch coverage, with a lower threshold (maybe `10-20%`), to ensure new changes are reasonably covered? Just wanted to check if this aligns with the intended behavior.

**Comment by felixarntz** [comment]
@swissspidy Since you raised some valid points in https://github.com/WordPress/performance/issues/1792#issuecomment-2630440480, it would be great to get your feedback whether this solution would address them.

File: codecov.yml (hunk 6 above)
Comment (swissspidy) [reply codecov.yml]: > I'm wondering if this was intentional—perhaps it was meant to ensure a minimum of `80%` coverage rather than allowing such a large drop?

I think so, yes. The naming is a bit confusing.

So maybe `threshold: 20%` or `target: 80%`?

**Comment by swissspidy** [comment]
@ShyamGadde Appreciate the thorough comparison here 💪 👍 

While individual flags per plugin are tempting, I think we can start with the simpler option 2 for now and then see how well it works.

File: codecov.yml [hunk 7]
```diff
@@ -8,8 +8,45 @@ coverage:
...
         informational: true
     patch:
       default:
-        threshold: 80%
+        threshold: 20%
         informational: true
+comment:
+  hide_comment_details: true
+  hide_project_coverage: false
+  show_carryforward_flags: true
```
Comment (ShyamGadde) [inline codecov.yml:21]: @westonruter Would you prefer that we display the carried forward flags in the PR comment, as shown below, or would you rather exclude them?

image

Ref: https://docs.codecov.com/docs/carryforward-flags#advanced-configuring-carryforward-flags-in-the-code-host

Comment (westonruter) [reply codecov.yml:21]: I suppose it doesn't hurt to include?

**Comment by phanduynam** [comment]
This issue has been resolved, please close it admin

**Comment by swissspidy** [comment]
@phanduynam No it's not, please stop spamming the repo with these comments.

File: .github/workflows/php-test-plugins.yml [hunk 8]
```diff
@@ -57,6 +59,55 @@ jobs:
     steps:
       - uses: styfle/cancel-workflow-action@0.12.1
       - uses: actions/checkout@v4
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
```
Comment (thelovekesh) [inline .github/workflows/php-test-plugins.yml]: This action was recently compromised. My recommendation would be to pin it to a specific SHA for better security.

Additionally, reviewing the workflow, I noticed there's quite a bit of scripting involved, which might not be the easiest to follow for all developers.

To improve clarity and flexibility, I'd suggest using a JavaScript or PHP script instead. This approach would make the code more accessible to a wider range of developers and allow for easier management of edge cases.

You can find a similar composite action here - https://github.com/ampproject/amp-wp/tree/develop/.github/actions/determine-changed-files

File: .github/workflows/php-test-plugins.yml [hunk 9]
```diff
@@ -88,48 +139,70 @@ jobs:
...
+              npm run test-php-multisite:$PLUGIN -- -- -- --coverage-clover=./multisite-reports/coverage-multisite-$PLUGIN.xml
+            done
           else
-            npm run test-php-multisite
+            for PLUGIN in ${{ steps.changed-plugins.outputs.all_changed_plugins }}; do
+              npm run test-php-multisite:$PLUGIN
+            done
           fi
-      - name: Upload single site coverage reports to Codecov
+      - name: Download Codecov CLI
```
Comment (thelovekesh) [inline .github/workflows/php-test-plugins.yml:166]: was there some issue with official action?

Comment (thelovekesh) [reply .github/workflows/php-test-plugins.yml:166]: ignore, I just read the comments above.

File: .github/workflows/php-test-plugins.yml (hunk 8 above)
Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: What if we pin it to a vulnerable version?

Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: But it seems like based on the latest supply chain attack with qix today, pinning is a way to avoid being exploited. Perhaps Dependabot includes some checks for the dependencies it is proposing updates for which would address the concern of pinning to a vulnerable version? It doesn't seem this is the case right now, but this would sure be a great use of AI.

Comment (thelovekesh) [reply .github/workflows/php-test-plugins.yml]: The best way to avoid these issues is to use your own solution rather than relying on third-party actions. It's generally a good idea to either use `@actions/*` or create your own composite actions.

Also, in core, it's very rare for maintainers to allow third-party actions, and when they do, they always pin to a specific SHA to mitigate potential security risks.

> What if we pin it to a vulnerable version?

By default, pin to the SHA of a version that was published at least a week ago. This provides ample time to detect if the action has been compromised.

Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: Thank you for the advice. I've opened a PR to implement this: https://github.com/WordPress/performance/pull/2171

Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: I just saw a mention of this action being compromised back in March: https://semgrep.dev/blog/2025/popular-github-action-tj-actionschanged-files-is-compromised/

This was around when the commits were being pushed to this branch. I suppose this means we should rotate the secrets? Granted, it has been 9 months.

File: codecov.yml [hunk 10]
```diff
@@ -8,8 +8,45 @@ coverage:
...
+flag_management:
+  default_rules:
+    carryforward: true
+  individual_flags:
+    - name: auto-sizes
+      paths:
+      - plugins/auto-sizes/
```
Comment (adamsilverstein) [inline codecov.yml]: the indenting on this line doesn't match the other plugins

Comment (westonruter) [reply codecov.yml]: Yeah. There is also a merge conflict due to #1605. So that merge conflict needs to be fixed, and then `npm run format-js` should fix up this YAML file as well.

File: .github/workflows/php-test-plugins.yml (hunk 8 above)
Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]:
```suggestion
        uses: tj-actions/changed-files@e0021407031f5be11a464abee9a0776171c79891 # v47.0.1
```

File: codecov.yml (hunk 10 above)
Comment (ShyamGadde) [reply codecov.yml]: Fixed it in bc5cfe9.

File: codecov.yml [hunk 11]
```diff
@@ -8,8 +8,45 @@ coverage:
...
+        - plugins/performance-lab/
+    - name: speculation-rules
+      paths:
+        - plugins/speculation-rules/
+    - name: web-worker-offloading
+      paths:
+        - plugins/web-worker-offloading/
+    - name: webp-uploads
+      paths:
+        - plugins/webp-uploads/
```
[the remark this answers is not in the document]
Comment (westonruter) [reply codecov.yml:55]: Added in 3a30efb1f48447398f7f9c7e927785b224a49936
